### PR TITLE
Add minimal authentication

### PR DESCRIPTION
**Pull request by quis**
**This is a proposal for what I think is the minimum we need to do to get authentication on the admin app**
image

To test this pull request, run the app after setting the following environment variable:
```
export DM_ADMIN_FRONTEND_PASSWORD_HASH='$p5k2$$salt$lHVacy6sSyQWKa953wuFNZwHD04CtzZb'
```
Then log in as `admin` with password `password`.

**This pull request may well not be enough because**
- I know very little about security and may have exposed some kind of glaring vulnerability here
- This just uses HTTP basic auth—if we want to do sessions/nice login screen then more work would be required

Comments encouraged.

### CODE REVIEW

**Comment by robyoung** [comment]
If all routes require authentication you could move the authentication to a before filter which would make it harder to accidentally forget when implementing a new view.
See: https://github.com/alphagov/digitalmarketplace-api/commit/167a05d31cd0a434b3a541c06cb2e7bd780b3c7a

**Comment by TheDoubleK** [comment]
I can't see a .htaccess file ;)

Otherwise, yep I reckon do as Rob says.

**Comment by quis** [comment]
@robyoung That's much neater. Thanks for the tip (and pre-written commit message): a430a5c

File: tests/app/test_application.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 class TestApplication(BaseApplicationTest):
     def test_index(self):
         response = self.client.get('/')
-        assert 200 == response.status_code
+        assert 401 == response.status_code
```
Comment (TheDoubleK) [inline tests/app/test_application.py]: We're going to need to work out how to log the tests in.  But maybe not today.

Comment (quis) [inline tests/app/test_application.py]: @TheDoubleK Probably have to disable authentication for the tests. And I'm not sure how we're going to deal with testing this app on production—don't want them editing real services.

File: app/main/__init__.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 from flask import Blueprint
+from .helpers.auth import requires_auth
 
 main = Blueprint('main', __name__)
 
+main.before_request(requires_auth)
```
Comment (TheDoubleK) [inline app/main/__init__.py]: Can this be wrapped in an ``if not test`` type thing?  There is the Test config that we can use. Think it should work.

Comment (quis) [inline app/main/__init__.py]: @TheDoubleK Figured it out this morning, have a look at 6e370fc and see if that meets your needs.

**Comment by TheDoubleK** [comment]
Nice.
Of course, we now have no way of unit-testing if the authentication works :/

**Pull request merged**